[DTensor] Added conversion from Replicate to _StridedShard

zpcore opened the pull request:
Add the conversion from `Replicate` to `_StridedShard`, similar to Shard.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171338
* #171219
* #171178
* #171023
* #170915
* #172266
* __->__ #171337

zpcore committed:
[DTensor] Added conversion from Replicate to _StridedShard

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"




[ghstack-poisoned]

XilunWu reviewed (commented):
What is the difference between this new method and `_shard_tensor(src_date_rank=None)`?

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"

Add the conversion from `Replicate` to `_StridedShard`, similar to Shard.




[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -832,6 +832,40 @@ def _to_replicate_tensor(
+            The local strided shard for this rank.
+        """
+        num_chunks = mesh.size(mesh_dim=mesh_dim)
+        shards, _ = self._split_tensor(
```
nit: invariably, every time i read one of the placement_types functions in a PR diff I am very confused becuase the diff does not show me which class is being edited, and the function names are all the same.  i could just as easily assume this method exists on the Replicate class as the StridedShard class.

I wonder if it would be worth a refactor where we extract these methods from the placement classes entirely and make them plain functions.  I noticed that 'self' is not used inside the definition of _split_tensor at all.  If we refactor, we could rename as well, and give more descriptive names

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"

Add the conversion from `Replicate` to `_StridedShard`, similar to Shard.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"

Add the conversion from `Replicate` to `_StridedShard`, similar to Shard.




[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -832,6 +832,40 @@ def _to_replicate_tensor(
+            The local strided shard for this rank.
+        """
+        num_chunks = mesh.size(mesh_dim=mesh_dim)
+        shards, _ = self._split_tensor(
```
this was addressed by the latest code change

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"

Add the conversion from `Replicate` to `_StridedShard`, similar to Shard.




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Added conversion from Replicate to _StridedShard"

Add the conversion from `Replicate` to `_StridedShard`, similar to Shard.




[ghstack-poisoned]

zpcore commented:
@pytorchbot merge
